dali/pipeline/operators/python_function/python_function.cc

```diff
@@ -115,17 +66,40 @@ void CopyNumpyArrayToTensor(Tensor<CPUBackend> &tensor, const py::array &array)
       static_cast<size_t>(contiguous.size() * contiguous.itemsize()));
 }
 
+py::list PrepareInputList(SampleWorkspace *ws, int idx) {
+  py::list args_list;
+  for (int i = 0; i < ws->NumInput(); ++i) {
+    auto &input = ws->Input<CPUBackend>(ws->NumInput() * idx + i);
+    py::dtype dtype(FormatStrFromType(input.type()));
+    auto input_array = py::array(dtype, input.shape(), input.raw_data(), py::array());
+    args_list.append(input_array);
+  }
+  return args_list;
+}
+
+void CopyOutputs(SampleWorkspace *ws, int idx, const py::tuple &output) {
+  for (int i = 0; i < ws->NumOutput(); ++i) {
+    auto &output_tensor = ws->Output<CPUBackend>(ws->NumInput() * idx + i);
+    auto output_array = py::cast<py::array>(output[i]);
+    CopyNumpyArrayToTensor(output_tensor, output_array);
+  }
+}
+
 template<>
 void PythonFunctionImpl<CPUBackend>::RunImpl(SampleWorkspace *ws, const int idx) {
-  const auto &input = ws->Input<CPUBackend>(idx);
-  auto &output = ws->Output<CPUBackend>(idx);
   py::gil_scoped_acquire guard{};
+  py::list args_list = PrepareInputList(ws, idx);
+  py::object output_o;
   try {
-    py::array output_array = python_function(TensorToNumpyArray(input));
-    CopyNumpyArrayToTensor(output, output_array);
+    output_o = python_function(*py::tuple(args_list));
   } catch(const py::error_already_set & e) {
     throw std::runtime_error(to_string("PythonFunction error: ") + to_string(e.what()));
   }
+  py::tuple output = (py::tuple::check_(output_o)) ? output_o : py::make_tuple(output_o);
+  DALI_ENFORCE(output.size() == static_cast<size_t>(ws->NumOutput()),
+               "Python function returned " + std::to_string(output.size()) + " outputs and "
+                   + std::to_string(ws->NumOutput()) + " were expected.");
```

Comment:
Maybe add test for this in `test_python_function_operator.py` ?

Reply:
Added.